httpclient/timeout_context.go

```diff
@@ -0,0 +1,120 @@
+package httpclient
+
+import (
+	"context"
+	"io"
+	"sync"
+	"time"
+)
+
+type timeoutContext struct {
+	ctx    context.Context
+	cancel context.CancelCauseFunc
```

Comment:
We should aim to deprecate go 1.19 at some point. It is currently EOL for 5 months. Will do a bit of datascience to see how many users would be affected by this deprecation. In the meantime, can we use `WithCancel()`?